README.md
# fluent-plugin-remote_syslog

[![Build Status](https://travis-ci.org/dlackty/fluent-plugin-remote_syslog.svg?branch=master)](https://travis-ci.org/dlackty/fluent-plugin-remote_syslog)

[Fluentd](http://fluentd.org) plugin for output to remote syslog serivce (e.g. [Papertrail](http://papertrailapp.com/))

## Requirements

| fluent-plugin-remote_syslog | fluentd                 | ruby   |
| -------------------         | ---------               | ------ |
| >= 1.0.0                    | >= v0.14.0 or >= v1.0.0 | >= 2.1 |
| < 1.0.0                     | >= v0.12.0              | >= 1.9 |

## Installation

* Fluentd

```bash
$ fluent-gem install fluent-plugin-remote_syslog
```

* td-agent

```bash
$ td-agent-gem install fluent-plugin-remote_syslog
```

## Usage

```
<match foo.bar>
  @type remote_syslog
  host example.com
  port 514
  severity debug
  program fluentd
  hostname ${tag[1]}

  <buffer tag>
  </buffer>

  <format>
    @type single_value
    message_key message
  </format>
</match>
```

## Configuration

| name              | type                             | placeholder support | description                                           |
| --------------    | -------                          | -----------         | ---------------------------------                     |
| hostname          | string                           | support             | departure of log                                      |
| host              | string                           | support             | syslog target host                                    |
| port              | integer (default: `514`)         |                     | syslog target port                                    |
| host_with_port    | string                           | support             | parameter for <host>:<port> style                     |
| facility          | string (default: `"user"`)       | support             | syslog facility                                       |
| severity          | string (default: `"notice"`      | support             | syslog severity                                       |
| program           | string (default: `"fluentd"`     | support             | syslog program name                                   |
| protocol          | enum (udp, tcp) (default: `udp`) |                     | transfer protocol                                     |
| tls               | bool (default: false)            |                     | use TLS (tcp only)                                    |
| ca_file           | string                           |                     | ca_file path (tls mode only)                          |
| verify_mode       | integer                          |                     | SSL verification mode (tls mode only)                 |
| packet_size       | integer (default: `1024`)        |                     | size limitation for syslog packet                     |
| timeout           | integer                          |                     | TCP transfer timeout. if value is 0, wait forever     |
| timeout_exception | bool (default: `false`)          |                     | if value is true, raise exception by transfer timeout |
| keep_alive        | bool (default: `false`)          |                     | use TCP keep alive                                    |
| keep_alive_idle   | integer                          |                     | set TCP keep alive idle time                          |
| keep_alive_cnt    | integer                          |                     | set TCP keep alive probe count                        |
| keep_alive_intvl  | integer                          |                     | set TCP keep alive probe interval                     |

### Common Configuration

#### Buffer Section

| name                        | default  |
| --------------              | -------  |
| flush_mode                  | interval |
| flush_interval              | 5        |
| flush_thread_interval       | 0.5      |
| flush_thread_burst_interval | 0.5      |

#### Format Section

| name           | default |
| -------------- | ------- |
| @type          | ltsv    |

## License

Copyright (c) 2014-2017 Richard Lee.  
Copyright (c) 2022 Daijiro Fukuda.

See LICENSE for details.


CHANGELOG.md
## 1.1.0

* Up rake dependency version for CVE-2020-8130
* Support "warning" Severity in addition to "warn" [#51](https://github.com/fluent-plugins-nursery/fluent-plugin-remote_syslog/pull/51)
* Make `packet_size` option valid to UDP [#46](https://github.com/fluent-plugins-nursery/fluent-plugin-remote_syslog/pull/46)

## 1.0.0

* Support fluentd-0.14 or later features [#24](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/24)
* Support TCP and TLS protocol [#24](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/24)
* Not support fluentd-0.12.0 from this version

## 0.3.3

* Allow override "severity" or "facility" from records [#19](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/19)
* Change default port to 514 [#15](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/15)

## 0.3.2

* Rewrite plugin to make rewrite tag function work properly [#4](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/4)

## 0.3.1

* Fix errors in last release [#3](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/3)

## 0.3.0 (yanked)

* Integrate with `Fluent::Mixin::RewriteTagName` [#2](https://github.com/dlackty/fluent-plugin-remote_syslog/pull/2)

## 0.2.1

* Fix encoding issue

## 0.2.0

* Integrate with `Fluent::Mixin::PlainTextFormatter`
* **BREAKING**: Remove `key_name` config, use `output_data_type` instead

## 0.1.0

* Initial release


test/test_helper.rb
$LOAD_PATH.unshift(File.join(File.dirname(__FILE__), "..", "lib"))
$LOAD_PATH.unshift(File.dirname(__FILE__))

require "minitest/autorun"
require "minitest/unit"
require "minitest/pride"
require "fluent/test"


test/plugin/out_remote_syslog.rb
require "test_helper"
require "fluent/plugin/out_remote_syslog"
require "fluent/plugin/in_syslog"
require 'fluent/test/driver/input'

class RemoteSyslogOutputTest < Test::Unit::TestCase
  def setup
    Fluent::Test.setup
  end

  def create_driver_out_syslog(conf)
    Fluent::Test::Driver::Output.new(Fluent::Plugin::RemoteSyslogOutput).configure(conf)
  end

  def create_driver_in_syslog(conf)
    Fluent::Test::Driver::Input.new(Fluent::Plugin::SyslogInput).configure(conf)
  end

  def test_configure
    d = create_driver_out_syslog %[
      @type remote_syslog
      hostname foo.com
      host example.com
      port 5566
      severity debug
      program minitest
    ]

    loggers = d.instance.instance_variable_get(:@senders)
    assert_equal loggers, []

    assert_equal "example.com", d.instance.instance_variable_get(:@host)
    assert_equal 5566, d.instance.instance_variable_get(:@port)
    assert_equal "debug", d.instance.instance_variable_get(:@severity)
  end

  data("debug", {severity: "debug"})
  data("warn", {severity: "warn"})
  data("warning", {severity: "warning", expected_severity: "warn"})
  def test_write_udp(data)
    out_driver = create_driver_out_syslog %[
      @type remote_syslog
      hostname foo.com
      host localhost
      port 5566
      severity #{data[:severity]}
      program minitest

      <format>
        @type single_value
        message_key message
      </format>
    ]

    in_driver = create_driver_in_syslog %[
      @type syslog
      tag tag
      port 5566
      bind 127.0.0.1
    ]

    in_driver.run(expect_records: 1, timeout: 5) do
      out_driver.run do
        out_driver.feed("tag", Fluent::EventTime.now, {"message" => "This is a test message."})
      end
    end

    assert_equal 1, in_driver.events.length

    tag, _, msg = in_driver.events[0]

    expected_facility = "user"
    expected_severity = data[:expected_severity] || data[:severity]
    assert_equal("tag.#{expected_facility}.#{expected_severity}", tag)
    assert_equal("foo.com", msg["host"])
    assert_equal("minitest", msg["ident"])
    assert_equal("This is a test message.", msg["message"])
  end

  data("debug", {severity: "debug"})
  data("warn", {severity: "warn"})
  data("warning", {severity: "warning", expected_severity: "warn"})
  def test_write_tcp(data)
    out_driver = create_driver_out_syslog %[
      @type remote_syslog
      hostname foo.com
      host localhost
      port 5566
      severity #{data[:severity]}
      program minitest

      protocol tcp

      <format>
        @type single_value
        message_key message
      </format>
    ]

    in_driver = create_driver_in_syslog %[
      @type syslog
      tag tag
      port 5566
      bind 127.0.0.1
      protocol_type tcp
    ]

    in_driver.run(expect_records: 1, timeout: 5) do
      out_driver.run do
        out_driver.feed("tag", Fluent::EventTime.now, {"message" => "This is a test message."})
      end
    end

    assert_equal 1, in_driver.events.length

    tag, _, msg = in_driver.events[0]

    expected_facility = "user"
    expected_severity = data[:expected_severity] || data[:severity]
    assert_equal("tag.#{expected_facility}.#{expected_severity}", tag)
    assert_equal("foo.com", msg["host"])
    assert_equal("minitest", msg["ident"])
    assert_equal("This is a test message.", msg["message"])
  end

  data("emerg", {in: "emerg", out: "emerg"})
  data("alert", {in: "alert", out: "alert"})
  data("crit", {in: "crit", out: "crit"})
  data("err", {in: "err", out: "err"})
  data("warn", {in: "warn", out: "warn"})
  data("warning", {in: "warning", out: "warn"})
  data("notice", {in: "notice", out: "notice"})
  data("info", {in: "info", out: "info"})
  data("debug", {in: "debug", out: "debug"})
  data("wrong", {in: "wrong", out: "wrong"})
  def test_severity_mapper(data)
    out = Fluent::Plugin::RemoteSyslogOutput::SeverityMapper.map(data[:in])
    assert_equal(data[:out], out)
  end
end


lib/fluent/plugin/out_remote_syslog.rb
require "remote_syslog_sender"

module Fluent
  module Plugin
    class RemoteSyslogOutput < Output
      Fluent::Plugin.register_output("remote_syslog", self)

      helpers :formatter, :inject

      config_param :hostname, :string, :default => ""

      config_param :host, :string, :default => nil
      config_param :port, :integer, :default => 514
      config_param :host_with_port, :string, :default => nil

      config_param :facility, :string, :default => "user"
      config_param :severity, :string, :default => "notice"
      config_param :program, :string, :default => "fluentd"

      config_param :protocol, :enum, list: [:udp, :tcp], :default => :udp
      config_param :tls, :bool, :default => false
      config_param :ca_file, :string, :default => nil
      config_param :verify_mode, :integer, default: nil
      config_param :packet_size, :size, default: 1024
      config_param :timeout, :time, default: nil
      config_param :timeout_exception, :bool, default: false

      config_param :keep_alive, :bool, :default => false
      config_param :keep_alive_idle, :integer, :default => nil
      config_param :keep_alive_cnt, :integer, :default => nil
      config_param :keep_alive_intvl, :integer, :default => nil

      config_section :buffer do
        config_set_default :flush_mode, :interval
        config_set_default :flush_interval, 5
        config_set_default :flush_thread_interval, 0.5
        config_set_default :flush_thread_burst_interval, 0.5
      end

      config_section :format do
        config_set_default :@type, 'ltsv'
      end

      def initialize
        super
      end

      def configure(conf)
        super
        if @host.nil? && @host_with_port.nil?
          raise ConfigError, "host or host_with_port is required"
        end

        @formatter = formatter_create
        unless @formatter.formatter_type == :text_per_line
          raise ConfigError, "formatter_type must be text_per_line formatter"
        end

        validate_target = "host=#{@host}/host_with_port=#{@host_with_port}/hostname=#{@hostname}/facility=#{@facility}/severity=#{@severity}/program=#{@program}"
        placeholder_validate!(:remote_syslog, validate_target)
        @senders = []
      end

      def multi_workers_ready?
        true
      end

      def close
        super
        @senders.each { |s| s.close if s }
        @senders.clear
      end

      def format(tag, time, record)
        r = inject_values_to_record(tag, time, record)
        @formatter.format(tag, time, r)
      end

      def write(chunk)
        return if chunk.empty?

        host = extract_placeholders(@host, chunk.metadata)
        port = @port

        if @host_with_port
          host, port = extract_placeholders(@host_with_port, chunk.metadata).split(":")
        end

        host_with_port = "#{host}:#{port}"

        Thread.current[host_with_port] ||= create_sender(host, port)
        sender = Thread.current[host_with_port]

        facility = extract_placeholders(@facility, chunk.metadata)
        severity = extract_placeholders(@severity, chunk.metadata)
        program = extract_placeholders(@program, chunk.metadata)
        hostname = extract_placeholders(@hostname, chunk.metadata)

        severity = SeverityMapper.map(severity)

        packet_options = {facility: facility, severity: severity, program: program}
        packet_options[:hostname] = hostname unless hostname.empty?

        begin
          chunk.open do |io|
            io.each_line do |msg|
              sender.transmit(msg.chomp!, packet_options)
            end
          end
        rescue
          if Thread.current[host_with_port]
            Thread.current[host_with_port].close
            @senders.delete(Thread.current[host_with_port])
            Thread.current[host_with_port] = nil
          end
          raise
        end
      end

      private

      def create_sender(host, port)
        if @protocol == :tcp
          options = {
            tls: @tls,
            whinyerrors: true,
            packet_size: @packet_size,
            timeout: @timeout,
            timeout_exception: @timeout_exception,
            keep_alive: @keep_alive,
            keep_alive_idle: @keep_alive_idle,
            keep_alive_cnt: @keep_alive_cnt,
            keep_alive_intvl: @keep_alive_intvl,
            program: @program,
          }
          options[:ca_file] = @ca_file if @ca_file
          options[:verify_mode] = @verify_mode if @verify_mode
          sender = RemoteSyslogSender::TcpSender.new(
            host,
            port,
            options
          )
        else
          sender = RemoteSyslogSender::UdpSender.new(
            host,
            port,
            whinyerrors: true,
            program: @program,
            packet_size: @packet_size,
          )
        end
        @senders << sender
        sender
      end

      # Convert some Severity values that is not supported in `syslog_protocol` library:
      #   https://github.com/eric/syslog_protocol
      # We want to fix `syslog_protocol` itself, but it seems to be not maintained for a while.
      # If the following PR is merged, we can remove this implementaion.
      #   https://github.com/eric/syslog_protocol/pull/9
      module SeverityMapper
        DICT = {
          # "warning" is not supported, but we should use it since "warn" is deprecated.
          "warning" => "warn", 
        }

        def self.map(severity)
          DICT[severity] || severity
        end
      end
    end
  end
end
